Native division for UInt128 on 32bit platforms

martinholters opened the pull request:
I'm sure there are clever tricks to pull off, but it seems to be an improvement against master already, and this not quite my field of expertise. Note that these benchmarks are made with a 32bit-julia (ARCH=i686) on a 64bit system.

Master:
```julia
julia> @btime $(one(UInt128)) ÷ $(one(UInt128));
  757.851 ns (11 allocations: 156 bytes)

julia> @btime $(-one(UInt128)) ÷ $(one(UInt128));
  868.255 ns (12 allocations: 180 bytes)

julia> @btime $(one(UInt128)) % $(one(UInt128));
  296.552 ns (8 allocations: 60 bytes)

julia> @btime $(-one(UInt128)) % $(one(UInt128));
  318.302 ns (8 allocations: 72 bytes)

julia> @eval Base function div(x::UInt128, y::UInt128)
           return UInt128(div(BigInt(x), BigInt(y)))::UInt128
       end;
       
julia> one(UInt128) ÷ one(UInt128);

julia> @btime $(one(UInt128)) ÷ $(one(UInt128));
  344.019 ns (10 allocations: 124 bytes)

julia> @btime $(-one(UInt128)) ÷ $(one(UInt128));
  451.320 ns (11 allocations: 148 bytes)

julia> @btime $(one(UInt128)) % $(one(UInt128));
  332.386 ns (10 allocations: 124 bytes)

julia> @btime $(-one(UInt128)) % $(one(UInt128));
  355.200 ns (10 allocations: 136 bytes)
```
The recursion limiting heuristics make it so that the results depend on which function is compiled first, so re-evaluating the original definition yields different results.

This PR:
```julia
julia> @btime $(one(UInt128)) ÷ $(one(UInt128));
  0.628 ns (0 allocations: 0 bytes)

julia> @btime $(-one(UInt128)) ÷ $(one(UInt128));
  383.621 ns (0 allocations: 0 bytes)

julia> @btime $(one(UInt128)) % $(one(UInt128));
  0.628 ns (0 allocations: 0 bytes)

julia> @btime $(-one(UInt128)) % $(one(UInt128));
  301.164 ns (0 allocations: 0 bytes)
```

I'm not 100% sure `0xfff...fff % 0x000...001` shows the worst-case performance, so there might be inputs where current master is slightly faster, but in any case, we get rid of the allocations.

martinholters committed:
Native division for UInt128 on 32bit platforms

By not relying on BigInts, the new code does not allocate and is, for 
many inputs, substantially faster. Also this fixes issues with the 
inferability.

martinholters committed:
Further simplify `rem(::UInt128, ::UInt128)`

...by not needlessly progressing in batches of 32.

martinholters commented:
The slight simplification of `rem` gives a 10% improvement here:
```julia
julia> @btime $(-one(UInt128)) % $(one(UInt128));
  277.050 ns (0 allocations: 0 bytes)
```

martinholters commented:
@simonbyrne, any thoughts on this? Anyone else willing to look at this? (I've tried a few minor modifications, none of which seemed to be a clear improvement.)

simonbyrne commented:
Sorry, this is probably a bit too far out of my expertise for me to review beyond "this would be good to have".

StefanKarpinski commented:
Do we have enough tests of 128-bit `divrem` to be confident that this is correct? If so, I think we can go ahead and merge this. Your call, @martinholters.

stevengj commented:
Do you have any reference for the algorithm you are using that you can add as a comment to the code?

The code is too hairy for me to easily proofread, so my main concern is that we have test coverage of all of the branches here.

staticfloat commented:
From a slightly different point of view, I like this because it decreases our dependence on GMP/MPFR, which helps for restrictive environments that don't allow those packages due to licensing reasons.

staticfloat commented:
Well, as far as correctness goes, it looks pretty good to me:

```
# <paste function here as `divrem2()`>

@testset "Random integers" begin
    for idx in 1:100000000
        x, y = rand(UInt128, 2)
        @test divrem2(x, y) == Base.divrem(x,y)
    end
end
```

That test took about 5 minutes to run, and it passed all cases, so that makes me pretty happy.

martinholters commented:
Yes, I have made similar randomized tests and they look good. However, getting good test coverage of the possible branches is pretty tough. E.g. I uncovered some bugs only by explicitly using small (random) divisors.

I have an idea for an algorithm which might be a bit easier to grok and first tests indicate that performance might be at least as good, but I'm not sure I'll find time to pursue this before January. But then again, there is no hurry here.

stevengj commented:
You'd need to explicitly design tests to cover the corner cases of small divisors etc.

martinholters committed:
Simplify `div(::UInt128, ::UInt128)` by introducing a `_setbit` helper

Processing in chunks of 32 bits was required for fast manipulation of 
`q`, which can be achieved with the optimized but-setting function in a 
cleaner way. The slight restrcutring of the remaining code is reflected 
in `rem(::UInt128, ::UInt128)`, too.

martinholters committed:
Add tests for UInt128 `div` and `rem`

martinholters commented:
>I have an idea for an algorithm which might be a bit easier to grok and first tests indicate that performance might be at least as good

Nope, turned out to be slower. Instead, I've tried to make the existing algorithm more readable. That called for introducing a function for setting the `i`-th bit in a `UInt128`. (`x |= UInt128(1) << i` leads to vastly sub-optimal code for the 32bit target.) I've named it `_setbit` for now, limited it to `UInt128`, and it's only defined on 32bit targets. Would it make sense to give it an underscore-free name and (unconditionally) provide the fallback `x | one(typeof(x)) << i` definition?

Also, I've added some tests.
